src/main/om/next/impl/parser.cljc

```diff
@@ -12,14 +12,15 @@
    A query expression is composed of EDN values. The grammar for query
    expressions follows:
 
-   QueryRoot    := EdnVector(QueryExpr*)
-   QueryExpr    := (EdnKeyword | IdentExpr | ParamExpr | JoinExpr)
-   IdentExpr    := EdnVector2(Keyword, EdnValue)
-   ParamExpr    := EdnList2(QueryExpr | EdnSymbol, ParamMapExpr)
-   ParamMapExpr := EdnMap(Keyword, EdnValue)
-   JoinExpr     := EdnMap((Keyword | IdentExpr), (QueryRoot | UnionExpr | RecurExpr))
-   UnionExpr    := EdnMap(Keyword, QueryRoot)
-   RecurExpr    := ('... | Integer)
+   QueryRoot      := EdnVector(QueryExpr*)
+   PlainQueryExpr := (EdnKeyword | IdentExpr | JoinExpr)
+   QueryExpr      := (PlainQueryExpr | ParamExpr)
+   IdentExpr      := EdnVector2(Keyword, EdnValue)
+   ParamExpr      := EdnList2(PlainQueryExpr | EdnSymbol, ParamMapExpr)
```

Comment:
A `ParamExpr` can't actually contain a `QueryExpr` because that would mean accepting a nested `ParamExpr` there. Change that to `PlainQueryExpr`.